12-offer-encoding.md

```diff
@@ -0,0 +1,1057 @@
+   * which are only valid if started January 1st 2021, or
+   * which if paid after January 1st you (over) pay the full rate first year, or
+   * which if paid after January 1st are paid pro-rata for the first year, or
+   * which repeat from whenever you made the first payment
+
+Thus, each payment has:
+1. A `time_unit` defining 0 (seconds), 1 (days), 2 (months), 3 (years).
+2. A `period`, defining how often (in `time_unit`) it has to be paid.
+3. An optional `recurrence_limit` of total payments to be paid.
+4. An optional `recurrence_base`:
+   * `basetime`, defining when the first period starts
+      in seconds since 1970-01-01 UTC.
+   * `start_any_period` if non-zero, meaning you don't have to start
+      paying at the period indicated by `basetime`, but can use
+      `recurrence_start` to indicate what period you are starting at.
+5. An optional `recurrence_paywindow`:
+   * `seconds_before`, defining how many seconds prior to the start of
+      the period a payment will be accepted.
+   * `proportional_amount`, if set indicating that a payment made during
+      the period itself will be charged proportionally to the remaining time
+      in the period (e.g. 150 seconds into a 1500 second period gives a 10%
+      discount).
+   * `seconds_after`, defining how many seconds after the start of the period
+      a payment will be accepted.
+  If this field is missing, payment will be accepted during the prior period and
+  the paid-for period.
+
+Note that the `absolute_expiry` field already covers the case where an offer
+is no longer valid after January 1st 2021.
+
+## Offer Period Calculation
+
+Each period has a zero-based index, and a start time and an end time.
+Because the periods can be in non-seconds units, the duration of a
+period can depend on when it starts.  The period with index N+1 begins
+immediately following the end of period with index N.
+
+- if an offer contains `recurrence_base`:
+  - the start of period #0 is `basetime` seconds since 1970-01-01 UTC.
+- otherwise:
+  - the start of period #0 is the time of issuance of the first
+    `invoice` for this particular offer and `payer_key`.
+
+To calculate the start of period #`N` for `N` > 0:
+- if `time_unit` is 0:
+  - period `N` starts at period #0 start plus `period` multiplied by `N`,
+    in seconds.
+- otherwise, if `time_unit` is 1:
+  - calculate the offset in seconds within the day of period #0 start.
+  - add `period` multiplied by `N` days to get the day of the period start.
+  - add the offset in seconds to get the period end in seconds.
+- otherwise, if `time_unit` is 2:
+  - calculate the offset in days within the month of period #0 start.
+  - calculate the offset in seconds within the day of period #0 start.
+  - add `period` multiplied by `N` months to get the month of the period start.
+  - add the offset days to get the day of the period start.
+    - if the day is not within the month, use the last day within the month.
```

Comment:
Shouldn't we also change the offset seconds to be a full day? If I start a one-month period at 04:00 on the 31st, it should last until 23:59 on the 30th of the next month (assuming it's a 30 days month), not until 04:00 on the 30th.
It feels unfair that a one-month period can end sooner than another one-month period that started earlier.